gdal/frmts/netcdf/netcdfsgwriterutil.h

```diff
@@ -68,67 +72,333 @@ namespace nccfdriver
+             * Directly subtracts the specified size from used_mem
+             */
+             void subCount(unsigned long long memfree) { this->used_mem -= memfree; }
+             unsigned long long& getUsage() { return used_mem; }
+
+             void reset() { this->used_mem = 0; }
+
+             WBuffer() {}
+    };
+
+
+
+    /* OGR_SGFS_Transaction
+     * Abstract class for a commitable transaction
      *
      */
-    class OGR_SGeometry_Scribe
+    class OGR_SGFS_Transaction
+    {
+        int varId = INVALID_VAR_ID;
+
+        public:
+            /* int commit(...);
+             * Arguments: int ncid, the dataset to write to
+             *            int write_loc, the index in which to write to
+             * Implementation: should write the transaction to netCDF file
+             *
+             */
+            virtual void commit(netCDFVID& n, size_t write_loc) = 0;
+
+            /* unsigned long long count(...)
+             * Implementation: supposed to return an approximate count of memory usage
+             * Most classes will implement with sizeof(*this), except if otherwise uncounted for dynamic allocation is involved.
+             */
+            virtual unsigned long long count() = 0;
+
+            /* appendToLog
+             * Implementation - given a file pointer, a transaction will be written to that log file in the format:
+             * -
+             * transactionVarId - sizeof(int) bytes
+             * NC_TYPE - sizeof(int) bytes
+             * (nc_char only) OP - 1 byte (0 if does not require COUNT or non-zero i.e. 1 if does)
+             * (nc_char only): SIZE of data - sizeof(size_t) bytes
+             * (nc_char && OP != 0 only) COUNT - sizeof(size_t) bytes the second dimension of the "count" arg
+             * DATA - size depends on NC_TYPE
+             */
+            virtual void appendToLog(VSILFILE*) = 0;
+
+            /* ~OGR_SGFS_Transaction()
+             * Empty. Simply here to stop the compiler from complaining...
+             */
+            virtual ~OGR_SGFS_Transaction() {}
+
+
+            /* OGR_SGFS_Transaction()
+             * Empty. Simply here to stop one of the CI machines from complaining...
+             */
+            OGR_SGFS_Transaction() {}
+
+            /* void getVarId(...);
+             * Gets the var in which to commit the transaction to.
+             */
+            int getVarId() { return this->varId; }
+
+            /* nc_type getType
+             * Returns the type of transaction being saved
+             */
+            virtual nc_type getType() = 0;
+
+            /* void setVarId(...);
+             * Sets the var in which to commit the transaction to.
+             */
+            void setVarId(int vId) { this->varId = vId; }
+
+    };
+
+    typedef std::map<int, void*> NCWMap;
+    typedef std::pair<int, void*> NCWEntry; // NC Writer Entry
+    typedef std::shared_ptr<OGR_SGFS_Transaction> MTPtr; // a.k.a Managed Transaction Ptr
```

Comment:
couldn't that be a unique_ptr ? That is are there several objects who can store the same MTPtr ?

Reply:
Yes to unique_ptr and no to the non-uniqueness property. Done :+1: .